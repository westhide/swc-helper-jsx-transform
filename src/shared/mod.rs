pub use transform::Transform;

mod transform;
