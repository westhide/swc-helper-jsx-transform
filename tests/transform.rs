#![feature(box_syntax)]
#![feature(box_patterns)]

use insta::assert_debug_snapshot;
use swc_core::ecma::{
    ast::Expr,
    parser::{Syntax, TsConfig},
    transforms::testing::Tester,
    visit::{as_folder, VisitMut},
};
use swc_helper_jsx_transform::shared::Transform;

const TSX_SYNTAX: Syntax = Syntax::Typescript(TsConfig {
    tsx: true,
    decorators: false,
    dts: false,
    no_early_errors: false,
});

macro_rules! snapshot {
    ($name:ident, $value:expr) => {
        assert_debug_snapshot!(stringify!($name), $value)
    };
}

macro_rules! test {
    ($name:ident, $src:literal) => {
        #[test]
        fn $name() {
            struct TransformVisitor;

            impl VisitMut for TransformVisitor {
                fn visit_mut_expr(&mut self, expr: &mut Expr) {
                    match expr {
                        Expr::JSXElement(box element) => {
                            snapshot!($name, element.transform())
                        },
                        Expr::JSXFragment(fragment) => {
                            snapshot!($name, fragment.transform())
                        },
                        _ => {},
                    }
                }
            }

            Tester::run(|tester| {
                tester.apply_transform(as_folder(TransformVisitor), "test.tsx", TSX_SYNTAX, $src)
            });
        }
    };
}

test!(tag_native_div, r#"<div></div>"#);
test!(tag_extra_component, r#"<A></A>"#);
test!(tag_extra_custom, r#"<tag-custom></tag-custom>"#);
test!(tag_extra_member, r#"<A.b></A.b>"#);

test!(attr_key_attr, r#"<div class="cls"></div>"#);
test!(attr_key_event_prefix_on, r#"<div onClick="()=>{}"></div>"#);
test!(
    attr_key_event_namespace_on,
    r#"<div on:click="()=>{}"></div>"#
);
test!(attr_key_namespace_attr, r#"<div ns:name="value"></div>"#);

test!(attr_value_lit, r#"<div class="1"></div>"#);
test!(
    attr_value_const_undefined,
    r#"<div class={undefined}></div>"#
);
test!(attr_value_const_array, r#"<div class={[1,2]}></div>"#);
test!(attr_value_const_obj, r#"<div class={{a:1}}></div>"#);
test!(attr_value_expr, r#"<div class={a}></div>"#);
test!(attr_value_element, r#"<div element=<div></div>></div>"#);
test!(
    attr_value_element_container,
    r#"<div element={<div></div>}></div>"#
);
test!(attr_value_fragment, r#"<div fragment=<></>></div>"#);
test!(
    attr_value_fragment_container,
    r#"<div fragment={<></>}></div>"#
);
test!(attr_value_empty, r#"<div class></div>"#);

test!(attr_spread, r#"<div {...spread}></div>"#);
test!(attr_boolean_attribute, r#"<div checked></div>"#);

test!(
    element_is_static,
    r#"
    <div class="cls1" style="background-color: red;">
      <div id="firstChild"></div>
    </div>
    "#
);

test!(
    fragment_with_children,
    r#"
    <>
      <div></div>
      <div></div>
    </>
    "#
);

test!(
    text_clean,
    r#"<div>
            text1
            <br/>
      text2
            </div>
    "#
);
