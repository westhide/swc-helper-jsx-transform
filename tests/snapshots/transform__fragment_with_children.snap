---
source: tests/transform.rs
expression: fragment.transform()
---
Fragment {
    children: [
        Element(
            Element {
                tag: Native(
                    "div",
                ),
                attrs: [],
                children: [],
                raw: JSXElement {
                    span: Span {
                        lo: BytePos(
                            15,
                        ),
                        hi: BytePos(
                            26,
                        ),
                        ctxt: #0,
                    },
                    opening: JSXOpeningElement {
                        name: Ident(
                            Ident {
                                span: Span {
                                    lo: BytePos(
                                        16,
                                    ),
                                    hi: BytePos(
                                        19,
                                    ),
                                    ctxt: #0,
                                },
                                sym: Atom('div' type=static),
                                optional: false,
                            },
                        ),
                        span: Span {
                            lo: BytePos(
                                15,
                            ),
                            hi: BytePos(
                                20,
                            ),
                            ctxt: #0,
                        },
                        attrs: [],
                        self_closing: false,
                        type_args: None,
                    },
                    children: [],
                    closing: Some(
                        JSXClosingElement {
                            span: Span {
                                lo: BytePos(
                                    20,
                                ),
                                hi: BytePos(
                                    26,
                                ),
                                ctxt: #0,
                            },
                            name: Ident(
                                Ident {
                                    span: Span {
                                        lo: BytePos(
                                            22,
                                        ),
                                        hi: BytePos(
                                            25,
                                        ),
                                        ctxt: #0,
                                    },
                                    sym: Atom('div' type=static),
                                    optional: false,
                                },
                            ),
                        },
                    ),
                },
                is_static: true,
            },
        ),
        Element(
            Element {
                tag: Native(
                    "div",
                ),
                attrs: [],
                children: [],
                raw: JSXElement {
                    span: Span {
                        lo: BytePos(
                            33,
                        ),
                        hi: BytePos(
                            44,
                        ),
                        ctxt: #0,
                    },
                    opening: JSXOpeningElement {
                        name: Ident(
                            Ident {
                                span: Span {
                                    lo: BytePos(
                                        34,
                                    ),
                                    hi: BytePos(
                                        37,
                                    ),
                                    ctxt: #0,
                                },
                                sym: Atom('div' type=static),
                                optional: false,
                            },
                        ),
                        span: Span {
                            lo: BytePos(
                                33,
                            ),
                            hi: BytePos(
                                38,
                            ),
                            ctxt: #0,
                        },
                        attrs: [],
                        self_closing: false,
                        type_args: None,
                    },
                    children: [],
                    closing: Some(
                        JSXClosingElement {
                            span: Span {
                                lo: BytePos(
                                    38,
                                ),
                                hi: BytePos(
                                    44,
                                ),
                                ctxt: #0,
                            },
                            name: Ident(
                                Ident {
                                    span: Span {
                                        lo: BytePos(
                                            40,
                                        ),
                                        hi: BytePos(
                                            43,
                                        ),
                                        ctxt: #0,
                                    },
                                    sym: Atom('div' type=static),
                                    optional: false,
                                },
                            ),
                        },
                    ),
                },
                is_static: true,
            },
        ),
    ],
    is_static: true,
}
